employee-rostering-frontend/src/ui/components/DataTable.test.tsx

```diff
@@ -17,16 +17,20 @@ import { mount, shallow } from 'enzyme';
 import toJson from 'enzyme-to-json';
 import * as React from 'react';
 import { DataTable, DataTableProps } from './DataTable';
-import { EditableComponent } from './EditableComponent';
+import { stringSorter} from 'util/CommonSorters';
+import FilterComponent from './FilterComponent';
 
 interface MockData {name: string; number: number}
 class MockDataTable extends DataTable<MockData, DataTableProps<MockData>> {
-  createNewDataInstance = jest.fn(() => ({name: '', number: 0}));
   displayDataRow = jest.fn((data) => [<span key={0} id="viewer">{data.name}</span>,
     <span key={1}>{data.number}</span>]);
+  getInitialStateForNewRow = jest.fn(() => ({}));
   editDataRow = jest.fn((data) => [<input key={0} id="editor" />,
     <input key={1} />]);
+  // @ts-ignore
```

Comment:
It seems this a result of `isValid()` mixing responsibilities. It does two different things:
1. it checks that edited value contains valid data and
2. it is a type predicate asserting that `editedValue: ReadonlyPartial<T>` is in fact a complete `T`.

Could you split these two functions into:
- `isT(editedValue: ReadonlyPartial<T>): editedValue is T;` and
- `isValid(data: T): boolean`?

Reply:
I was thinking of refactoring the Validation logic where each column has its own list of validators (which allow us to display feedback on the input component causing the error). So I will probably add a `isDataComplete(editedValue: ReadonlyPartial<T>): editedValue is T;` and a `getValidatorLists(): Validator[][]` (two-dimensional since each column can have multiple validators with different error messages for each one).

Reply:
Did a refactoring of validation logic (not pushed yet). I used `formsy-react` (https://github.com/formsy/formsy-react) (2.0 beta) and `validator` (https://github.com/validatorjs/validator.js) to validate the row and provide feedback. This is done by wrapping validated inputs into a `Validated` element, which attaches the validators.

Reply:
Yes. Per-column validation sounds like a good idea, although I'd says it's the next level. Could you do this in a separate PR (unless it actually reduces the PR size :slightly_smiling_face:)?